components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1269,6 +1270,255 @@ where
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+        );
+        assert_eq!(self.fsm.peer.get_role(), StateRole::Candidate);
+
+        if let Some(ForceLeaderState::PreForceLeader {
+            expected_alive_voter,
+        }) = self.fsm.peer.force_leader.take()
+        {
+            let peer_ids: Vec<_> = self
+                .fsm
+                .peer
+                .raft_group
+                .raft
+                .prs()
+                .conf()
+                .voters()
+                .ids()
+                .iter()
+                .collect();
+            for peer_id in peer_ids {
+                let store_id = self
+                    .region()
+                    .get_peers()
+                    .iter()
+                    .find(|p| p.get_id() == peer_id)
+                    .unwrap()
+                    .get_store_id();
+
+                if expected_alive_voter.contains(&store_id) {
+                    continue;
+                }
+
+                // make fake vote response
+                let mut msg = raft::eraftpb::Message::new();
+                msg.msg_type = MessageType::MsgRequestVoteResponse;
+                msg.reject = false;
+                msg.term = self.fsm.peer.term();
+                msg.from = peer_id;
+                msg.to = self.fsm.peer.peer_id();
+                self.fsm.peer.raft_group.step(msg).unwrap();
+            }
+            // after receiving all votes, should become leader
+            assert!(self.fsm.peer.is_leader());
+            self.fsm.peer.raft_group.raft.set_check_quorum(false);
+
+            // make sure it's not hibernated
+            self.reset_raft_tick(GroupState::Ordered);
+
+            self.fsm.peer.force_leader = Some(ForceLeaderState::ForceLeader {
+                expected_alive_voter,
+            });
+        } else {
+            unreachable!();
+        }
+        self.fsm.has_ready = true;
+    }
+
+    fn on_exit_force_leader(&mut self) {
+        if self.fsm.peer.force_leader.is_none() {
+            return;
+        }
+
+        info!(
+            "exit force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+        );
+        self.fsm.peer.force_leader = None;
+        // make sure it's not hibernated
+        self.reset_raft_tick(GroupState::Ordered);
+        // leader lease shouldn't be renewed in force leader state.
+        assert_eq!(
+            self.fsm.peer.leader_lease.inspect(None),
+            LeaseState::Expired
+        );
+        self.fsm
+            .peer
+            .raft_group
+            .raft
+            .become_follower(self.fsm.peer.term() + 1, raft::INVALID_ID);
+
+        self.fsm.peer.raft_group.raft.set_check_quorum(true);
+        self.fsm.peer.raft_group.raft.pre_vote = true;
+        // let it trigger election immediately.
+        let _ = self.fsm.peer.raft_group.campaign();
```

Comment:
What if it's still learner after being recovered?

Reply:
Even promote the learner to voter or allow learner become leader, if the learner happens in a tiflash node, it may be required to generate snapshot in SST format, not sure if it is supported currently

Reply:
@BusyJay PD ensures it will be promoted as a voter, though better check `promotable` here. If not, do not campaign. will change

Reply:
@NingLin-P It's not supported yet. For the tiflash peer which has the latest log, PD will trigger remove the learner locally and not let it be the leader. I'm supplementing the issues about merge and tiflash in design doc, and let's discuss it then.